GH-3183: Add ReactiveRequestHandlerAdvice (#3197)
spring-projects/spring-integration: src/reference/asciidoc/handler-advice.adoc

----
====

[[reactive-advice]]
==== Reactive Advice

Starting with version 5.3, a `ReactiveRequestHandlerAdvice` can be used for request message handlers producing a `Mono` replies.
A `BiFunction<Message<?>, Mono<?>, Publisher<?>>` has to be provided for this advice and it is called from the `Mono.transform()` operator on a reply produced by the intercepted `handleRequestMessage()` method implementation.
Typically such a `Mono` customization is necessary when we would like to control network fluctuations via `timeout()`, `retry()` and similar support operators.

> For example when we can an HTTP request over WebFlux client, we could use bellow configuration to not wait for response more than 5 seconds:

Reviewer: ```suggestion
For example when we can an HTTP request over WebFlux client, we could use below configuration to not wait for response more than 5 seconds:
```

====
[source, java]
----
.handle(WebFlux.outboundGateway("https://somehost/"),
                       e -> e.customizeMonoReply((message, mono) -> mono.timeout(Duration.ofSeconds(5))));
----
====

> The `message` argument is a request message for such a message handler and can be used to determine a request-scope attributes.

Reviewer: ```suggestion
The `message` argument is the request message for the message handler and can be used to determine request-scope attributes.
```

> The `mono` argument is a result of this message handler `handleRequestMessage()` method implementation.

Reviewer: ```suggestion
The `mono` argument is the result of this message handler's `handleRequestMessage()` method implementation.
```
